examples/dnsrewrite/common.lua

```diff
@@ -0,0 +1,220 @@
+local linux = require("linux")
+local string = require("string")
+
+local common = {}
+
+local udp = 0x11
+local dns = 0x35  -- DNS port 53
+local eth_len = 14
+
+-- Target domain: test.internal
+local target_dns = string.pack("s1s1", "test", "internal")
+
+-- Target IP: 127.0.0.1
+local target_ip = 0x7F000001  -- 127.0.0.1 in network byte order (big-endian)
+
+local function get_domain(skb, off)
+	local _, nameoff, name = skb:getstring(off):find("([^\0]*)")
+	return name, nameoff + 1
+end
+
+local function calculate_ip_checksum(skb, ip_off, ip_len)
+	-- Zero out the checksum field first
+	skb:setuint16(ip_off + 10, 0)
+
+	local sum = 0
+	-- Sum all 16-bit words in the IP header
+	for i = 0, ip_len - 1, 2 do
+		sum = sum + skb:getuint16(ip_off + i)
+	end
+
+	-- Fold 32-bit sum to 16 bits
+	while (sum >> 16) > 0 do
+		sum = (sum & 0xFFFF) + (sum >> 16)
+	end
+
+	return ~sum & 0xFFFF
+end
+
+local function calculate_udp_checksum(skb, ip_off, ihl, udp_off, udp_len)
+	-- Get source and destination IP addresses
+	local src_ip = skb:getuint32(ip_off + 12)
+	local dst_ip = skb:getuint32(ip_off + 16)
+
+	-- Zero out the UDP checksum field first
+	skb:setuint16(udp_off + 6, 0)
+
+	local sum = 0
+
+	-- Add pseudo-header
+	sum = sum + ((src_ip >> 16) & 0xFFFF)
+	sum = sum + (src_ip & 0xFFFF)
+	sum = sum + ((dst_ip >> 16) & 0xFFFF)
+	sum = sum + (dst_ip & 0xFFFF)
+	sum = sum + udp  -- Protocol
+	sum = sum + udp_len  -- UDP length
+
+	-- Add UDP header and data
+	for i = 0, udp_len - 1, 2 do
+		sum = sum + skb:getuint16(udp_off + i)
+	end
+
+	-- Fold 32-bit sum to 16 bits
+	while (sum >> 16) > 0 do
+		sum = (sum & 0xFFFF) + (sum >> 16)
+	end
+
+	local checksum = ~sum & 0xFFFF
+	-- UDP checksum of 0x0000 should be sent as 0xFFFF
+	if checksum == 0 then
+		checksum = 0xFFFF
+	end
+
+	return checksum
+end
+
+function common.hook(skb, action)
+	-- Get protocol from IP header (offset 9)
+	local proto = skb:getuint8(eth_len + 9)
+
+	-- Only process UDP packets
+	if proto ~= udp then
+		return action.ACCEPT
+	end
+
+	-- Get IP header length (IHL field in first byte of IP header)
+	local ihl = skb:getuint8(eth_len) & 0x0F
+	local thoff = eth_len + ihl * 4  -- Transport header offset
+
+	-- Check if destination port is DNS (53) - this means it's a DNS query
+	local dstport = linux.ntoh16(skb:getuint16(thoff + 2))
+	if dstport ~= dns then
```

Comment:
we can decide this at nft level as well.. (not mandatory though)

Reply:
I too think handling these at nft level will be better, it will skip the lua C API calls.